### PR TITLE
Fix issue where its possible for a component to receive a unit without a config

### LINKED ISSUE #2086: Beats managed by agent can be left without a configured output
When the Beats restart after an output configuration has changed, we've observed that starting in 8.6 some of them can be left in a failed state with an `output unit has no config` error. Example of the unit state from a recent agent diagnostics bundle:

```yaml
  state:
    state: 2
    message: 'Healthy: communicating with pid ''5565'''
    units:
      ? unittype: 0
        unitid: beat/metrics-monitoring-metrics-monitoring-beats
      : state: 4
        message: '[output unit has no config]'
      ? unittype: 1
        unitid: beat/metrics-monitoring
      : state: 4
        message: '[output unit has no config]'
    version_info:
      name: beat-v2-client
      version: 8.6.0
      meta:
        build_time: 2023-01-04 01:30:07 +0000 UTC
        commit: 561a3e1839f1a50ce832e8e114de399b2bee2542
```

So far reports of this have primarily affected the Beats started for use in internal monitoring, but I have not been able to confirm if this bug exclusively affects the monitoring Beats.

We need to reproduce this and fix it. It appears to be triggered by the Beats restarting when the output changes. A similar problem was observed in https://github.com/elastic/beats/issues/34137 which was fixed in https://github.com/elastic/elastic-agent/pull/2036. There must be another problem with unit configurations after a Beat restarts.

Relevant log snippet of the full state transition:

```json
{"log.level":"info","@timestamp":"2023-01-11T19:21:11.997Z","message":"beat is restarting because output changed","component":{"binary":"metricbeat","dataset":"elastic_agent.metricbeat","id":"beat/metrics-monitoring","type":"beat/metrics"},"log.logger":"centralmgmt.V2-manager","log.origin":{"file.line":503,"file.name":"management/managerV2.go"},"service.name":"metricbeat","ecs.version":"1.6.0","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2023-01-11T19:21:11.998Z","log.origin":{"file.name":"coordinator/coordinator.go","file.line":835},"message":"Unit state changed beat/metrics-monitoring (HEALTHY->STOPPING): Restarting","component":{"id":"beat/metrics-monitoring","state":"HEALTHY"},"unit":{"id":"beat/metrics-monitoring","type":"output","state":"STOPPING","old_state":"HEALTHY"},"ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2023-01-11T19:21:11.999Z","log.origin":{"file.name":"coordinator/coordinator.go","file.line":835},"message":"Unit state changed beat/metrics-monitoring-metrics-monitoring-beats (CONFIGURING->STOPPING): Stopping","component":{"id":"beat/metrics-monitoring","state":"HEALTHY"},"unit":{"id":"beat/metrics-monitoring-metrics-monitoring-beats","type":"input","state":"STOPPING","old_state":"CONFIGURING"},"ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2023-01-11T19:21:12.004Z","message":"Uptime: 1.228625389s","component":{"binary":"metricbeat","dataset":"elastic_agent.metricbeat","id":"beat/metrics-monitoring","type":"beat/metrics"},"ecs.version":"1.6.0","log.logger":"monitoring","log.origin":{"file.line":196,"file.name":"log/log.go"},"service.name":"metricbeat","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2023-01-11T19:21:12.004Z","message":"metricbeat stopped.","component":{"binary":"metricbeat","dataset":"elastic_agent.metricbeat","id":"beat/metrics-monitoring","type":"beat/metrics"},"service.name":"metricbeat","ecs.version":"1.6.0","log.origin":{"file.line":491,"file.name":"instance/beat.go"},"ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2023-01-11T19:21:12.009Z","log.origin":{"file.name":"coordinator/coordinator.go","file.line":837},"message":"Component state changed beat/metrics-monitoring (HEALTHY->STOPPED): Suppressing FAILED state due to restart for '5536' exited with code '0'","component":{"id":"beat/metrics-monitoring","state":"STOPPED","old_state":"HEALTHY"},"ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2023-01-11T19:21:12.009Z","log.origin":{"file.name":"coordinator/coordinator.go","file.line":837},"message":"Unit state changed beat/metrics-monitoring (STOPPING->STOPPED): Suppressing FAILED state due to restart for '5536' exited with code '0'","component":{"id":"beat/metrics-monitoring","state":"STOPPED"},"unit":{"id":"beat/metrics-monitoring","type":"output","state":"STOPPED","old_state":"STOPPING"},"ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2023-01-11T19:21:12.009Z","log.origin":{"file.name":"coordinator/coordinator.go","file.line":837},"message":"Unit state changed beat/metrics-monitoring-metrics-monitoring-beats (STOPPING->STOPPED): Suppressing FAILED state due to restart for '5536' exited with code '0'","component":{"id":"beat/metrics-monitoring","state":"STOPPED"},"unit":{"id":"beat/metrics-monitoring-metrics-monitoring-beats","type":"input","state":"STOPPED","old_state":"STOPPING"},"ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2023-01-11T19:21:13.011Z","log.origin":{"file.name":"coordinator/coordinator.go","file.line":825},"message":"Spawned new component beat/metrics-monitoring: Starting: spawned pid '5565'","component":{"id":"beat/metrics-monitoring","state":"STARTING"},"ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2023-01-11T19:21:13.011Z","log.origin":{"file.name":"coordinator/coordinator.go","file.line":825},"message":"Spawned new unit beat/metrics-monitoring: Starting: spawned pid '5565'","component":{"id":"beat/metrics-monitoring","state":"STARTING"},"unit":{"id":"beat/metrics-monitoring","type":"output","state":"STARTING"},"ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2023-01-11T19:21:13.011Z","log.origin":{"file.name":"coordinator/coordinator.go","file.line":825},"message":"Spawned new unit beat/metrics-monitoring-metrics-monitoring-beats: Starting: spawned pid '5565'","component":{"id":"beat/metrics-monitoring","state":"STARTING"},"unit":{"id":"beat/metrics-monitoring-metrics-monitoring-beats","type":"input","state":"STARTING"},"ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2023-01-11T19:21:13.180Z","message":"Output reload is enabled, the beat will restart as needed on change of output config","component":{"binary":"metricbeat","dataset":"elastic_agent.metricbeat","id":"beat/metrics-monitoring","type":"beat/metrics"},"log.logger":"centralmgmt","log.origin":{"file.line":119,"file.name":"management/managerV2.go"},"service.name":"metricbeat","ecs.version":"1.6.0","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2023-01-11T19:21:13.180Z","message":"running under elastic-agent, per-beat lockfiles disabled","component":{"binary":"metricbeat","dataset":"elastic_agent.metricbeat","id":"beat/metrics-monitoring","type":"beat/metrics"},"log.origin":{"file.line":401,"file.name":"instance/beat.go"},"service.name":"metricbeat","ecs.version":"1.6.0","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2023-01-11T19:21:13.184Z","message":"Setup Beat: metricbeat; Version: 8.6.0","component":{"binary":"metricbeat","dataset":"elastic_agent.metricbeat","id":"beat/metrics-monitoring","type":"beat/metrics"},"service.name":"metricbeat","ecs.version":"1.6.0","log.origin":{"file.line":296,"file.name":"instance/beat.go"},"ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2023-01-11T19:21:13.188Z","message":"Output is configured through Central Management","component":{"binary":"metricbeat","dataset":"elastic_agent.metricbeat","id":"beat/metrics-monitoring","type":"beat/metrics"},"ecs.version":"1.6.0","log.origin":{"file.line":325,"file.name":"instance/beat.go"},"service.name":"metricbeat","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2023-01-11T19:21:13.224Z","message":"metricbeat start running.","component":{"binary":"metricbeat","dataset":"elastic_agent.metricbeat","id":"beat/metrics-monitoring","type":"beat/metrics"},"log.origin":{"file.line":486,"file.name":"instance/beat.go"},"service.name":"metricbeat","ecs.version":"1.6.0","ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2023-01-11T19:21:13.231Z","log.origin":{"file.name":"coordinator/coordinator.go","file.line":831},"message":"Component state changed beat/metrics-monitoring (STARTING->HEALTHY): Healthy: communicating with pid '5565'","component":{"id":"beat/metrics-monitoring","state":"HEALTHY","old_state":"STARTING"},"ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2023-01-11T19:21:13.294Z","log.origin":{"file.name":"coordinator/coordinator.go","file.line":827},"message":"Unit state changed beat/metrics-monitoring-metrics-monitoring-beats (STARTING->CONFIGURING): Configuring","component":{"id":"beat/metrics-monitoring","state":"HEALTHY"},"unit":{"id":"beat/metrics-monitoring-metrics-monitoring-beats","type":"input","state":"CONFIGURING","old_state":"STARTING"},"ecs.version":"1.6.0"}
{"log.level":"error","@timestamp":"2023-01-11T19:21:13.431Z","log.origin":{"file.name":"coordinator/coordinator.go","file.line":833},"message":"Unit state changed beat/metrics-monitoring (STARTING->FAILED): [output unit has no config]","component":{"id":"beat/metrics-monitoring","state":"HEALTHY"},"unit":{"id":"beat/metrics-monitoring","type":"output","state":"FAILED","old_state":"STARTING"},"ecs.version":"1.6.0"}
{"log.level":"error","@timestamp":"2023-01-11T19:21:13.431Z","log.origin":{"file.name":"coordinator/coordinator.go","file.line":833},"message":"Unit state changed beat/metrics-monitoring-metrics-monitoring-beats (CONFIGURING->FAILED): [output unit has no config]","component":{"id":"beat/metrics-monitoring","state":"HEALTHY"},"unit":{"id":"beat/metrics-monitoring-metrics-monitoring-beats","type":"input","state":"FAILED","old_state":"CONFIGURING"},"ecs.version":"1.6.0"}
{"log.level":"info","@timestamp":"2023-01-11T19:21:13.499Z","log.origin":{"file.name":"coordinator/coordinator.go","file.line":827},"message":"Unit state changed beat/metrics-monitoring-metrics-monitoring-beats (FAILED->CONFIGURING): Configuring","component":{"id":"beat/metrics-monitoring","state":"HEALTHY"},"unit":{"id":"beat/metrics-monitoring-metrics-monitoring-beats","type":"input","state":"CONFIGURING","old_state":"FAILED"},"ecs.version":"1.6.0"}
{"log.level":"error","@timestamp":"2023-01-11T19:21:13.601Z","log.origin":{"file.name":"coordinator/coordinator.go","file.line":833},"message":"Unit state changed beat/metrics-monitoring-metrics-monitoring-beats (CONFIGURING->FAILED): [output unit has no config]","component":{"id":"beat/metrics-monitoring","state":"HEALTHY"},"unit":{"id":"beat/metrics-monitoring-metrics-monitoring-beats","type":"input","state":"FAILED","old_state":"CONFIGURING"},"ecs.version":"1.6.0"}
```

Interestingly in these logs the non-monitoring Beats did not restart, only the monitoring Beats restarted.

### PR DESCRIPTION
**Pull request by blakerouse**
## What does this PR do?

Fixes an issue where it is possible for a component to receive a unit without a configuration. This was occurring because the usage of the checkinExpected channel was incorrect.

This fixes the issue by ensures that channel is empty on a new Checkin RPC call and prevent the sending goroutine from reading messages from the channel until the initial checkin has been processed.

This also adds a very defensive path where it only takes the latest `CheckinExpected` message and sends it, but only after parsing all previous messages to ensure that the latest message does not have a missing configuration that was sent in a previous `CheckinExpected`.

This also adds more testing.

## Why is it important?

It was possible for a component to restart with the a missing configuration for a unit, which prevents the component from working correctly and it reports itself as failed.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `./changelog/fragments` using the [changelog tool](https://github.com/elastic/elastic-agent#changelog)

## Related issues

- Closes #2086

### CODE REVIEW

File: pkg/component/runtime/manager_test.go [hunk 1]
```diff
@@ -1162,14 +1175,312 @@ LOOP:
...
+		Units: []component.Unit{
+			{
+				ID:       "fake-input",
+				Type:     client.UnitTypeInput,
+				LogLevel: client.UnitLogLevelTrace,
+				Config: component.MustExpectedConfig(map[string]interface{}{
+					"type":    "fake",
+					"state":   int(client.UnitStateHealthy),
+					"message": "Fake Healthy",
+					"killer":  true,
```
Comment (cmacknz) [inline pkg/component/runtime/manager_test.go:1376]: Either add a comment that this causes the component to randomly die, or rename the config option to obviously state this (like `randomlySelfTerminate` or something). I had to go look a the fake component implementation to remember what this is doing, and it will be harder to figure out without looking with the paired change there in the same diff.

File: pkg/component/runtime/manager_test.go [hunk 2]
```diff
@@ -1162,14 +1175,312 @@ LOOP:
...
+	}
+
+	subCancel()
+	cancel()
+
+	err = <-errCh
+	require.NoError(t, err)
+}
+
+func TestManager_FakeInput_KeepsRestarting(t *testing.T) {
```
Comment (cmacknz) [inline pkg/component/runtime/manager_test.go:1327]: Did this or the other test ever reproduce the problem? What about if you reverted my `ClearPendingCheckin` change that didn't completely fix the problem? Does it catch it then or not at all?

**Review by cmacknz — COMMENTED** [review]
On first pass this looks good. Will continue tomorrow, I haven't run the tests or installed the agent yet to double check this works.

File: pkg/component/runtime/manager_test.go (hunk 2 above)
Comment (cmacknz) [reply pkg/component/runtime/manager_test.go:1327]: Do the tests catch the issue with no fixes applied at all? (is what I was trying to ask above)

File: pkg/component/fake/shipper/listener.go [hunk 3]
```diff
@@ -21,6 +22,13 @@ func createListener(path string) (net.Listener, error) {
 	if _, err := os.Stat(path); !os.IsNotExist(err) {
 		os.Remove(path)
 	}
+	dir := filepath.Dir(path)
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply pkg/component/fake/shipper/listener.go:25]: There was a bug in the test that its possible that the host doesn't have `/tmp/elastic-agent` wasn't created. So this fixes that issue to ensure that the directory exists.

File: pkg/component/runtime/runtime_comm.go [hunk 4]
```diff
@@ -190,12 +221,24 @@ func (c *runtimeComm) checkin(server proto.ElasticAgent_CheckinV2Server, init *p
...
+		for {
+			select {
+			case <-checkinDone:
+				return
+			case <-recvDone:
+				return
+			case <-waitExp:
+				break WAIT
+			}
+		}
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply pkg/component/runtime/runtime_comm.go:241]: Added comment.

File: pkg/component/runtime/manager_test.go [hunk 5]
```diff
@@ -1162,14 +1175,312 @@ LOOP:
...
+		t.Fatalf("failed early: %s", err)
+	}
+
+	endTimer := time.NewTimer(1 * time.Minute)
+	defer endTimer.Stop()
+LOOP:
+	for {
+		select {
+		case <-endTimer.C:
+			t.Fatalf("timed out after 30 seconds")
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply pkg/component/runtime/manager_test.go:1311]: Fixed.

File: pkg/component/runtime/manager_test.go [hunk 6]
```diff
@@ -1162,14 +1175,312 @@ LOOP:
...
```
[the remark this answers is not in the document]
Comment (blakerouse) [reply pkg/component/runtime/manager_test.go:1460]: Fixed.

File: pkg/component/runtime/manager_test.go (hunk 1 above)
Comment (blakerouse) [reply pkg/component/runtime/manager_test.go:1376]: It's not random it is on an interval. I will work on a better name.

File: pkg/component/runtime/manager_test.go (hunk 2 above)
Comment (blakerouse) [reply pkg/component/runtime/manager_test.go:1327]: I just tried it with your change removed and sadly these did not show the symptoms either. I wanted to keep the tests in the change because overall I thought they still exercised the code path well.

Comment (cmacknz) [reply pkg/component/runtime/manager_test.go:1327]: Agreed I like them too, maybe if we ran them for long enough they'd catch it.

File: pkg/component/runtime/runtime_comm.go [hunk 7]
```diff
@@ -141,30 +137,65 @@ func (c *runtimeComm) CheckinExpected(expected *proto.CheckinExpected) {
...
-	default:
-	}
 }
 
 func (c *runtimeComm) CheckinObserved() <-chan *proto.CheckinObserved {
 	return c.checkinObserved
 }
 
+// latestCheckinExpected ensures that the latest expected checkin is used
+func (c *runtimeComm) latestCheckinExpected(exp *proto.CheckinExpected) *proto.CheckinExpected {
```
Comment (cmacknz) [inline pkg/component/runtime/runtime_comm.go:148]: This function could be unit tested by itself it looks like. Do we exercise any of the edge cases already? If not it is probably worth adding some sanity tests.

**Review by cmacknz — APPROVED** [review]
Ran the tests and confirmed the agent can install and enroll with Fleet.

It is difficult to tell if the issue is completely gone (I thought it was the first time) but these changes look like they should prevent it.

**Comment by blakerouse** [comment]
Using the latest changes in this branch I added the following change and ran the `TestManager_FakeInput_Restarts` test. It causes the exact same issue that we are seeing in production where the `Config` is `nil`.

So my change still does not 100% fix the issue, because it is possible that the goroutine reads from the channel and what is on the channel is invalid for a restart component.

I am working more on this PR to have a full proof solution. I want to ensure there is no window of possibility that this happens. I am also looking at adding some defensive code on the elastic-agent-client side to ensure that even if it does happen it handles the case correctly.

**Comment by cmacknz** [comment]
I completely removed all of the fixes including the first attempt at a fix that this PR replaces and ran the two unit tests here continuously for ~1 hour without reproducing a failure. 

This appears particularly challenging to reproduce in an isolated way, I'm glad we managed to recreate it artificially at least.

File: pkg/component/runtime/runtime_comm.go (hunk 7 above)
Comment (blakerouse) [reply pkg/component/runtime/runtime_comm.go:180]: Being that I have changed the implementation I was able to remove all this extra code. It is no longer required.

File: pkg/component/runtime/runtime_comm.go [hunk 8]
```diff
@@ -137,6 +145,30 @@ func (c *runtimeComm) CheckinExpected(expected *proto.CheckinExpected) {
...
 	}
+
+	// we need to determine if the communicator is currently in the initial observed message path
+	// in the case that it is we send the expected state over a different channel
+	c.initCheckinObservedMx.Lock()
+	initObserved := c.initCheckinObserved
+	expectedCh := c.initCheckinExpectedCh
+	if initObserved != nil {
+		// the next call to `CheckinExpected` must be from the initial `CheckinObserved` message
+		if observed != initObserved {
```
Comment (cmacknz) [inline pkg/component/runtime/runtime_comm.go:156]: Just to double check, we only care about equality of the references here? I know these are both pointers so I assume so.

If you need to compare the proto messages by value you need to use `proto.Equal(observed, initObserved)` otherwise this comparison will have unpredictable results. 

It _looks_ fine but we had a previous bug where we failed to use `proto.Equal` so I want to confirm the intent here.

File: pkg/component/runtime/runtime_comm.go [hunk 9]
```diff
@@ -137,6 +145,30 @@ func (c *runtimeComm) CheckinExpected(expected *proto.CheckinExpected) {
...
+			// not the initial observed message; we don't send it
```
Comment (cmacknz) [inline pkg/component/runtime/runtime_comm.go:157]: Is there any scenario where the initial observed message could be lost or altered before we get here? Unlikely but if it this were possible we'd be stuck unable to configure anything.

Is there a client side timeout on the Checkin request from the component? If there is that would guard against this by giving it a way to make a second request and reset the initCheckinObserved message.

File: pkg/component/runtime/runtime_comm.go (hunk 8 above)
Comment (blakerouse) [reply pkg/component/runtime/runtime_comm.go:156]: Yes we only care about comparing the reference. We want to be sure that the observed message is the exact message that was the initial message, not just the same contents.

File: pkg/component/runtime/runtime_comm.go (hunk 9 above)
Comment (cmacknz) [reply pkg/component/runtime/runtime_comm.go:157]: If the lack of an expected message being sent to the client would cause it to miss subsequent checkins that would also get us out of this theoretical state via the eventual restart of the process.

**Comment by cmacknz** [comment]
The latest round of changes looks good to me, although I would feel better with some targeted tests to ensure this definitely fixes the bug, or that we can't accidentally deadlock the checkin process waiting for an initial message that doesn't match what we expect.

File: pkg/component/runtime/runtime_comm.go (hunk 9 above)
Comment (blakerouse) [reply pkg/component/runtime/runtime_comm.go:157]: Yes it will result in an eventual restart of the process by the command runtime. So this situation is self healing, but extremely unlikely.

**Comment by cmacknz** [comment]
Installed locally and changed the output configuration a few times without issue.

File: pkg/component/runtime/runtime_comm.go (hunk 9 above)
Comment (pkoutsovasilis) [reply pkg/component/runtime/runtime_comm.go:157]: @cmacknz @blakerouse  I need a sync with either of you (or both) to tell me more about the assumptions here, We have this issue https://github.com/elastic/elastic-agent/issues/7944 and this is getting a lot of traction lately as to why endpoint remains dormant 🙂

Comment (cmacknz) [reply pkg/component/runtime/runtime_comm.go:157]: > Yes it will result in an eventual restart of the process by the command runtime. So this situation is self healing, but extremely unlikely.

Reading this 2 years later, self-healing except for the service runtime...

I at least don't have most of the context of this PR top of mind anymore.
